File: scripts/limits.py
import argparse
import ast
from fnmatch import fnmatchcase
import io
import json
from pathlib import Path
import re
import subprocess
import sys
import tokenize


def command(args, cwd):
    return subprocess.check_output(args, cwd=cwd, text=True, stderr=subprocess.PIPE)


def checked_prose(text):
    output = []
    fence = None
    comment = False
    for line in text.splitlines():
        stripped = line.lstrip(' ')
        if len(line) - len(stripped) <= 3:
            if comment:
                if '-->' in stripped:
                    comment = False
                continue
            if stripped.startswith('<!--'):
                comment = '-->' not in stripped
                continue
            match = re.match(r'(`{3,}|~{3,})', stripped)
            if match:
                marker = match.group(1)
                if fence is None:
                    fence = marker
                elif marker[0] == fence[0] and len(marker) >= len(fence):
                    fence = None
                continue
        if fence is None:
            output.append(line)
    return '\n'.join(output)


DENIED_DIRECTORIES = {'atlas', 'data', 'state', 'intake', 'graph', 'plans', 'runs', 'secrets', '.claude', '.codex', '.agents'}
DENIED_FILES = {'*.sqlite*', '*.db*', '*.jsonl', '.env', '.env.*', 'engine.pin', 'copies-manifest', 'delivery-registry'}
REQUIRED_IGNORES = {name + '/' for name in DENIED_DIRECTORIES} | DENIED_FILES


def hygiene(root, tracked):
    errors = []
    def published(name):
        if name in tracked:
            return subprocess.check_output(['git', 'show', ':' + name], cwd=root)
        return (root / name).read_bytes()
    if '.gitignore' not in tracked:
        errors.append('.gitignore must be tracked')
    else:
        lines = set(published('.gitignore').decode().splitlines())
        for pattern in sorted(REQUIRED_IGNORES - lines):
            errors.append(f'.gitignore missing required pattern: {pattern}')
        if any(line.startswith('!') for line in lines):
            errors.append('.gitignore must not negate private-data ignores')
    candidates = tracked | set(filter(None, command(['git', 'ls-files', '--others', '--exclude-standard', '-z'], root).split('\0')))
    for name in sorted(candidates):
        if (root / name).is_symlink():
            errors.append(f"{name}: symlink visible to public Git layer")
            continue
        if any(part in DENIED_DIRECTORIES or any(fnmatchcase(part, pattern) for pattern in DENIED_FILES) for part in Path(name).parts):
            errors.append(f'private path visible to public Git layer: {name}')
        if name.startswith('fixtures/') and b'Vera Example' not in published(name):
            errors.append(f'{name}: fixture lacks Vera Example marker')
    return errors


def check(root):
    errors = []
    entries = command(['git', 'ls-files', '--stage', '-z'], root).split('\0')
    files = {}
    tracked = set()
    for entry in filter(None, entries):
        metadata, name = entry.split('\t', 1)
        mode, _, stage = metadata.split()
        tracked.add(name)
        if stage != '0':
            errors.append(f'{name}: unresolved index entry')
        if mode == '120000' or (root / name).is_symlink():
            errors.append(f'{name}: tracked symlink is forbidden')
            continue
        if mode == '160000':
            continue
        try:
            files[name] = (root / name).read_bytes()
        except OSError as error:
            errors.append(f'{name}: {error}')
    errors.extend(hygiene(root, tracked))
    budget = sum(len(data) for name, data in files.items() if name != 'LICENSE' and Path(name).name != 'Cargo.lock') / 4
    if budget > 70_000:
        errors.append(f'budget exceeded: {budget:g} > 70000 tokens')
    directories = set()
    for name in files:
        parts = Path(name).parts
        if not any(part.startswith('.') for part in parts):
            directories.update('/'.join(parts[:i]) + '/' for i in range(1, len(parts)))
        if name.lower().endswith('.md') and name not in {'README.md', 'AGENTS.md', 'GOALS.md', 'CLAUDE.md'}:
            errors.append(f'{name}: Markdown is not allowed')
    readme = files.get('README.md', b'').decode()
    section = re.search(r'^## Map\s*\n(.*?)(?=^#|\Z)', readme, re.M | re.S)
    mapped = []
    if section is None:
        errors.append('README.md: missing ## Map')
    else:
        for line in section.group(1).splitlines():
            if not line.strip():
                continue
            match = re.fullmatch(r'- `([^`]+/)`: .+', line)
            if match is None or len(line) > 250:
                errors.append(f'README.md: invalid Map line: {line}')
            else:
                mapped.append(match.group(1))
        if sorted(mapped) != sorted(directories):
            errors.append(f'README.md: Map must list each visible directory exactly once: {sorted(directories)}')
    metadata = json.loads(command(['cargo', 'metadata', '--offline', '--no-deps', '--format-version', '1', '--manifest-path', str(root / 'Cargo.toml')], root))
    package = next(p for p in metadata['packages'] if Path(p['manifest_path']).resolve() == (root / 'Cargo.toml').resolve())
    target_names = {str(Path(t['src_path']).relative_to(root)): t['name'] for t in package['targets'] if 'test' in t['kind']}
    targets = set(target_names)
    if any(Path(name).parent != Path('tests') or not name.endswith('.rs') for name in targets):
        errors.append('Rust integration tests must live in tests/*.rs')
    checker = Path(__file__).resolve().parents[1] / 'Cargo.toml'
    sources = [{'path': name, 'text': data.decode(), 'test_target': name in targets} for name, data in files.items() if name.endswith('.rs')]
    result = subprocess.run(['cargo', 'run', '--quiet', '--locked', '--manifest-path', str(checker), '--example', 'check_limits'], input=json.dumps(sources), text=True, capture_output=True, check=True)
    rust = json.loads(result.stdout)
    errors.extend(rust['errors'])
    if set(rust['test_files']) != targets:
        errors.append('Rust test targets must all be tracked')
    goals = checked_prose(files.get('GOALS.md', b'').decode())
    bindings = []
    for paragraph in re.split(r'\n\s*\n', goals):
        for match in re.finditer(r'^\d+\. .*(?:\n[ \t]+.*)*', paragraph, re.M):
            names = re.findall(r'`([^`]+\.rs)`', match.group())
            if len(names) != 1:
                errors.append('GOALS.md: each numbered goal must name exactly one Rust test file')
            bindings.extend(names)
    if sorted(bindings) != sorted(targets):
        errors.append(f'GOALS.md: bind each Rust test target exactly once: {sorted(targets)}')
    for name, data in files.items():
        if not name.endswith('.py'):
            continue
        try:
            source = data.decode()
            for token in tokenize.generate_tokens(io.StringIO(source).readline):
                if token.type == tokenize.COMMENT and not (token.start[0] == 1 and token.string.startswith('#!')):
                    errors.append(f'{name}:{token.start[0]}: Python comments are forbidden')
            for node in ast.walk(ast.parse(source)):
                if isinstance(node, (ast.Module, ast.FunctionDef, ast.AsyncFunctionDef, ast.ClassDef)) and ast.get_docstring(node) is not None:
                    errors.append(f'{name}: Python docstrings are forbidden')
        except (SyntaxError, tokenize.TokenError) as error:
            errors.append(f'{name}: invalid Python: {error}')
    if not errors:
        for path, name in target_names.items():
            listing = command(['cargo', 'test', '--offline', '--locked', '--manifest-path', str(root / 'Cargo.toml'), '--test', name, '--', '--list'], root)
            if not re.search(r'^.+: test$', listing, re.M):
                errors.append(f'{path}: Cargo discovered no runnable tests')
    print(f'Budget: {budget:g}/70000 tokens (tracked bytes / 4; LICENSE and Cargo.lock excluded)')
    print('\n'.join(errors) if errors else 'Limits: OK')
    return bool(errors)


def main():
    parser = argparse.ArgumentParser(description='Atlas limits, with Cargo target discovery and Rust syntax checks')
    parser.add_argument('--root', type=Path, default=Path.cwd())
    args = parser.parse_args()
    try:
        return check(args.root.resolve())
    except (OSError, ValueError, StopIteration, subprocess.CalledProcessError) as error:
        print(f'limits failed: {error}', file=sys.stderr)
        if isinstance(error, subprocess.CalledProcessError):
            print(error.stderr, file=sys.stderr)
        return 1


if __name__ == '__main__':
    sys.exit(main())

File: tests/limits.rs
use std::fs;
use std::path::Path;
use std::process::{Command, Output};
use tempfile::TempDir;

fn limits(root: &Path) -> Output {
    Command::new(if cfg!(windows) { "python" } else { "python3" })
        .arg(Path::new(env!("CARGO_MANIFEST_DIR")).join("scripts/limits.py"))
        .arg("--root")
        .arg(root)
        .output()
        .unwrap()
}

fn stage(root: &Path) {
    assert!(
        Command::new("git")
            .current_dir(root)
            .args(["add", "--all", "--force", "--", ".", ":!target"])
            .status()
            .unwrap()
            .success()
    );
}

fn fixture() -> TempDir {
    let root = tempfile::Builder::new()
        .prefix("vera-example-limits-")
        .tempdir()
        .unwrap();
    assert!(
        Command::new("git")
            .current_dir(root.path())
            .args(["init", "--quiet"])
            .status()
            .unwrap()
            .success()
    );
    fs::create_dir(root.path().join("src")).unwrap();
    fs::create_dir(root.path().join("tests")).unwrap();
    for (name, text) in [
        (
            "Cargo.toml",
            "[package]\nname = \"vera-example-limits\"\nversion = \"0.1.0\"\nedition = \"2024\"\n",
        ),
        (
            "src/main.rs",
            "fn main() { let _ = \"Vera Example // literal\"; }\n",
        ),
        (
            "tests/check.rs",
            "#[test]\nfn vera_example() { let _ = r###\"Vera Example /* literal */\"###; }\n",
        ),
        (
            "README.md",
            "# Vera Example\n\n## Map\n\n- `src/`: Vera Example executable.\n- `tests/`: Vera Example tests.\n",
        ),
        (
            "GOALS.md",
            "# Vera Example\n\n1. Vera Example boundary. `tests/check.rs`\n",
        ),
        ("AGENTS.md", "Vera Example rules.\n"),
    ] {
        fs::write(root.path().join(name), text).unwrap();
    }
    fs::copy(
        Path::new(env!("CARGO_MANIFEST_DIR")).join(".gitignore"),
        root.path().join(".gitignore"),
    )
    .unwrap();
    assert!(
        Command::new("cargo")
            .current_dir(root.path())
            .args(["generate-lockfile", "--offline"])
            .status()
            .unwrap()
            .success()
    );
    stage(root.path());
    root
}

fn expect_failure(root: &Path, needle: &str) {
    let output = limits(root);
    let text = format!(
        "{}{}",
        String::from_utf8_lossy(&output.stdout),
        String::from_utf8_lossy(&output.stderr)
    );
    assert!(
        !output.status.success(),
        "checker accepted invalid fixture: {needle}"
    );
    assert!(text.contains(needle), "missing {needle}: {text}");
}

#[test]
fn repository_meets_its_limits() {
    let output = limits(Path::new(env!("CARGO_MANIFEST_DIR")));
    assert!(
        output.status.success(),
        "{}{}",
        String::from_utf8_lossy(&output.stdout),
        String::from_utf8_lossy(&output.stderr)
    );
}

#[test]
fn rust_checks_distinguish_literals_from_comments_and_reject_source_inclusion() {
    let root = fixture();
    let valid = limits(root.path());
    assert!(
        valid.status.success(),
        "{}{}",
        String::from_utf8_lossy(&valid.stdout),
        String::from_utf8_lossy(&valid.stderr)
    );
    for (source, message) in [
        (
            "// Vera Example comment\nfn main() {}\n",
            "Rust comments are forbidden",
        ),
        (
            "/* Vera Example block */ fn main() {}\n",
            "Rust comments are forbidden",
        ),
        (
            "/// Vera Example docs\nfn main() {}\n",
            "Rust comments are forbidden",
        ),
        (
            "#[doc = \"Vera Example docs\"]\nfn main() {}\n",
            "documentation attributes are forbidden",
        ),
        (
            "fn main() {}\n#[test] fn vera_example_internal() {}\n",
            "goal-bound integration targets",
        ),
        (
            "fn main() {}\n#[cfg(test)] mod checks {}\n",
            "test modules do not belong",
        ),
    ] {
        fs::write(root.path().join("src/main.rs"), source).unwrap();
        stage(root.path());
        expect_failure(root.path(), message);
    }
    fs::write(root.path().join("src/main.rs"), "fn main() {}\n").unwrap();
    for (source, message) in [
        (
            "include!(\"../src/main.rs\");\n#[test] fn vera_example() {}\n",
            "not source inclusion",
        ),
        (
            "#[path = \"../src/main.rs\"] mod implementation;\n#[test] fn vera_example() {}\n",
            "must not load source",
        ),
        (
            "#[test] #[ignore] fn vera_example() {}\n",
            "must not be ignored",
        ),
        (
            "use std::include as load; load!(\"../src/main.rs\"); #[test] fn vera_example() {}\n",
            "source-inclusion macros",
        ),
        (
            "#[cfg(any())] #[test] fn vera_example() {}\n",
            "Cargo discovered no runnable tests",
        ),
    ] {
        fs::write(root.path().join("tests/check.rs"), source).unwrap();
        stage(root.path());
        expect_failure(root.path(), message);
    }
    fs::write(
        root.path().join("tests/check.rs"),
        "fn vera_example_but_not_a_test() {}\n",
    )
    .unwrap();
    stage(root.path());
    expect_failure(root.path(), "has no #[test]");
}

#[test]
fn goals_map_budget_and_markdown_are_enforced() {
    for (name, contents, needle) in [
        (
            "GOALS.md",
            "# Vera Example\n\n1. Unbound Vera Example goal.\n",
            "exactly one Rust test",
        ),
        (
            "GOALS.md",
            "# Vera Example\n\n1. Vera Example. `tests/check.rs`\n\n2. Vera Example duplicate. `tests/check.rs`\n",
            "bind each Rust test target exactly once",
        ),
        (
            "README.md",
            "# Vera Example\n\n## Map\n\n- `src/`: Vera Example.\n",
            "Map must list",
        ),
        (
            "extra.md",
            "Vera Example extra Markdown\n",
            "Markdown is not allowed",
        ),
    ] {
        let root = fixture();
        fs::write(root.path().join(name), contents).unwrap();
        stage(root.path());
        expect_failure(root.path(), needle);
    }
    let root = fixture();
    fs::write(
        root.path().join("large.txt"),
        "Vera Example ".repeat(24_000),
    )
    .unwrap();
    stage(root.path());
    expect_failure(root.path(), "budget exceeded");
    let root = fixture();
    fs::write(
        root.path().join("README.md"),
        format!(
            "# Vera Example\n\n## Map\n\n- `src/`: {}\n- `tests/`: Vera Example.\n",
            "Vera Example ".repeat(25)
        ),
    )
    .unwrap();
    stage(root.path());
    expect_failure(root.path(), "invalid Map line");
}

#[test]
fn public_hygiene_uses_staged_content_and_rejects_private_paths() {
    let root = fixture();
    fs::create_dir(root.path().join("fixtures")).unwrap();
    fs::write(
        root.path().join("fixtures/sample.txt"),
        "Invented but unmarked",
    )
    .unwrap();
    stage(root.path());
    fs::write(
        root.path().join("fixtures/sample.txt"),
        "Vera Example unstaged masking edit",
    )
    .unwrap();
    expect_failure(root.path(), "fixture lacks Vera Example");
    for name in [
        "secrets/token.txt",
        "nested/example.sqlite-wal",
        "nested/.env.secret",
        "nested/run.jsonl",
    ] {
        let root = fixture();
        let path = root.path().join(name);
        fs::create_dir_all(path.parent().unwrap()).unwrap();
        fs::write(path, "Vera Example forbidden fixture").unwrap();
        stage(root.path());
        expect_failure(root.path(), "private path visible");
    }
    let root = fixture();
    let ignore = fs::read_to_string(root.path().join(".gitignore")).unwrap();
    fs::write(
        root.path().join(".gitignore"),
        ignore.replace("data/\n", ""),
    )
    .unwrap();
    stage(root.path());
    fs::write(root.path().join(".gitignore"), ignore).unwrap();
    expect_failure(root.path(), "missing required pattern: data/");
}

#[test]
#[cfg(unix)]
fn tracked_symlinks_fail() {
    let root = fixture();
    std::os::unix::fs::symlink("src/main.rs", root.path().join("vera-link")).unwrap();
    stage(root.path());
    expect_failure(root.path(), "tracked symlink is forbidden");
}
